server/src/main/java/org/elasticsearch/index/fieldvisitor/StoredFieldLoader.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.fieldvisitor;
+
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.lucene.index.SequentialStoredFieldsLeafReader;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Generates a {@link LeafStoredFieldLoader} for a given lucene segment to load stored fields.
+ */
+public abstract class StoredFieldLoader {
+
+    /**
+     * Return a {@link LeafStoredFieldLoader} for the given segment and document set
+     *
+     * The loader will use an internal lucene merge reader if the document set is of
+     * sufficient size and is contiguous.  Callers may pass {@code null} if the set
+     * is not known up front or if the merge reader optimisation will not apply.
+     */
+    public abstract LeafStoredFieldLoader getLoader(LeafReaderContext ctx, int[] docs);
+
+    /**
+     * @return a list of fields that will be loaded for each document
+     */
+    public abstract List<String> fieldsToLoad();
+
+    private static List<String> fieldsToLoad(boolean loadSource, Set<String> fields) {
+        Set<String> fieldsToLoad = new HashSet<>();
+        fieldsToLoad.add("_id");
+        fieldsToLoad.add("_routing");
+        if (loadSource) {
+            fieldsToLoad.add("_source");
+        }
+        fieldsToLoad.addAll(fields);
+        return fieldsToLoad.stream().sorted().toList();
+    }
```

Comment:
Maybe shove this at the end of the file so readers see the public interface first.